clients/src/main/java/org/apache/kafka/common/utils/Utils.java

```diff
@@ -1519,6 +1537,32 @@ public static String[] enumOptions(Class<? extends Enum<?>> enumClass) {
                 .toArray(String[]::new);
     }
 
+    /**
+     * Ensure that the class is concrete (i.e., not abstract). If it is, throw a {@link ConfigException}
+     * with a friendly error message suggesting a list of concrete child subclasses (if any are known).
+     * @param cls the class to check; may not be null
+     * @param name the name of the type of class to use in the error message; e.g., "Transform",
+     *             "Interceptor", or even just "Class"; may be null
+     * @throws ConfigException if the class is not concrete
+     */
+    public static void ensureConcrete(Class<?> cls, String name) {
+        Objects.requireNonNull(cls);
+        if (isBlank(name))
+            name = "Class";
+        if (Modifier.isAbstract(cls.getModifiers())) {
+            String childClassNames = Stream.of(cls.getClasses())
+                    .filter(cls::isAssignableFrom)
```

Comment:
I took a stab at something that would accept a base class and relax the filtering logic here accordingly, but it didn't quite feel right since we don't do verification that `cls` extends `baseClass`. I guess we could add that check as well and have something like `Utils::ensureConcreteSubclass`--how does that sound? I'm on the fence about it potentially doing too much for a utility method, but it does provide user-facing advantages.

Reply:
> I guess we could add that check as well and have something like Utils::ensureConcreteSubclass--how does that sound?

Yeah I imagine a `ensureConcreteSubclass(Class<?> cls, Class<?> baseClass)` that:
1. asserts that cls is concrete and baseClass.isAssignableFrom(cls) is true
2. if not valid, suggests child classes that are concrete and`baseClass::isAssignableFrom` is true.

I don't think that it would compose well with the current `ensureConcrete` method though, so it would probably replace what is there already.

> I'm on the fence about it potentially doing too much for a utility method, but it does provide user-facing advantages.

I thought that about the child class suggestion logic, but relented for the same reason. It significantly complicates this method which could otherwise be a single if condition, but that complication could help someone quickly resolve an obvious typo.

I don't think that asserting `baseClass.isAssignableFrom(cls)` is too much for this utility method, because we should always have a baseClass in mind when defining a `CLASS` configuration. Users of the `AbstractConfig::getConfiguredInstance` methods already need to provide a baseClass for subclass checking and type safety at runtime, this utility method would help users to remember the same check at validation time.

And worst-case, someone can just specify Object as the baseClass, and effectively disable the baseClass filtering.